tests/test_flow.py

```diff
@@ -644,6 +645,96 @@ def append_body_and_return(lst, x):
     assert termination_result == [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9]]
 
 
+def test_batch_by_user_key():
+    def append_and_return(lst, x):
+        lst.append(x)
+        return lst
+
+    controller = build_flow(
+        [
+            Source(),
+            Batch(2, 100, "value"),
+            Reduce([], lambda acc, x: append_and_return(acc, x)),
+        ]
+    ).run()
+
+    values_1 = [i for i in range(4)]
+    values_2 = [i for i in range(4)]
+    values_3 = [i for i in range(4)]
+    values_4 = [i for i in range(4)]
+
+    for _ in range(4):
+        rand_val_1 = choice(values_1)
```

Comment:
If everything is random you can't really validate specific flows, maybe better to have test specific flows with a specific predefined input

Reply:
Also better add tests for other scenarios like:
* timeout of a some keys.
* reaching max events of key2 while key1 timeout is "sleeping"
* and a bunch of other variations